apisix/plugins/elasticsearch-logging.lua

```diff
@@ -0,0 +1,154 @@
+
+local DEFAULT_ELASTICSEARCH_SOURCE = "apache-apisix-elasticsearch-logging"
+
+local plugin_name = "elasticsearch-logging"
+local batch_processor_manager = bp_manager_mod.new(plugin_name)
+local str_format = core.string.format
+local str_sub = string.sub
+
+
+local schema = {
+    type = "object",
+    properties = {
+        endpoint = {
+            type = "object",
+            properties = {
+                uri = core.schema.uri_def,
+                index = { type = "string"},
+                type = { type = "string"},
+                username = { type = "string"},
+                password = { type = "string"},
+                timeout = {
+                    type = "integer",
+                    minimum = 1,
+                    default = 10
+                },
+                ssl_verify = {
+                    type = "boolean",
+                    default = true
+                }
+            },
+            required = { "uri", "index" }
+        },
+    },
+    required = { "endpoint" },
+}
+
+
+local _M = {
+    version = 0.1,
+    priority = 413,
+    name = plugin_name,
+    schema = batch_processor_manager:wrap_schema(schema),
+}
+
+
+function _M.check_schema(conf)
+    return core.schema.check(schema, conf)
+end
+
+
+local function get_logger_entry(conf)
+    local entry = log_util.get_full_log(ngx, conf)
+    return core.json.encode({
+            create = {
+                _index = conf.endpoint.index,
+                _type = conf.endpoint.type
+            }
+        }) .. "\n" ..
+        core.json.encode({
+            time = ngx_now(),
+            host = entry.server.hostname,
+            source = DEFAULT_ELASTICSEARCH_SOURCE,
+            request_url = entry.request.url,
+            request_method = entry.request.method,
+            request_headers = entry.request.headers,
+            request_query = entry.request.querystring,
+            request_size = entry.request.size,
+            response_headers = entry.response.headers,
+            response_status = entry.response.status,
+            response_size = entry.response.size,
+            latency = entry.latency,
+            upstream = entry.upstream,
+        }) .. "\n"
+end
+
+
+local function send_to_elasticsearch(conf, entries)
+    local httpc, err = http.new()
+    if not httpc then
+        return false, str_format("create http error: %s", err)
+    end
+
+    local uri = conf.endpoint.uri ..
+        (str_sub(conf.endpoint.uri, -1) == "/" and "_bulk" or "/_bulk")
+    local body = core.table.concat(entries, "")
+    local headers = {["Content-Type"] = "application/json"}
+    if conf.endpoint.username and conf.endpoint.password then
+        local authorization = "Basic " .. ngx.encode_base64(
+            conf.endpoint.username .. ":" .. conf.endpoint.password
+        )
+        headers["Authorization"] = authorization
+    end
+
+    core.log.info("uri: ", uri, ", body: ", body, ", headers: ", core.json.encode(headers))
```

Comment:
This `headers` may contain `username` and `passowrd` in base64 format, which we should not output in the log for security reasons.